// @solid-devtools/debugger/dist/chunk-R4OAKKPG.js
// packages/debugger/src/inspector/types.ts
var INFINITY = "Infinity";
var NEGATIVE_INFINITY = "NegativeInfinity";
var NAN = "NaN";
var UNDEFINED = "undefined";
var ValueType = /* @__PURE__ */ ((ValueType2) => {
  ValueType2["Number"] = "number";
  ValueType2["Boolean"] = "boolean";
  ValueType2["String"] = "string";
  ValueType2["Null"] = "null";
  ValueType2["Symbol"] = "symbol";
  ValueType2["Array"] = "array";
  ValueType2["Object"] = "object";
  ValueType2["Function"] = "function";
  ValueType2["Getter"] = "getter";
  ValueType2["Element"] = "element";
  ValueType2["Instance"] = "instance";
  ValueType2["Store"] = "store";
  ValueType2["Unknown"] = "unknown";
  return ValueType2;
})(ValueType || {});
var PropGetterState = /* @__PURE__ */ ((PropGetterState2) => {
  PropGetterState2["Live"] = "live";
  PropGetterState2["Stale"] = "stale";
  return PropGetterState2;
})(PropGetterState || {});
var GLOBAL_GET_VALUE = "$SdtGetValue";

// packages/debugger/src/locator/types.ts
var WINDOW_PROJECTPATH_PROPERTY = "$sdt_projectPath";
var LOCATION_ATTRIBUTE_NAME = "data-source-loc";

// packages/debugger/src/main/constants.ts
var DevtoolsMainView = /* @__PURE__ */ ((DevtoolsMainView3) => {
  DevtoolsMainView3["Structure"] = "structure";
  return DevtoolsMainView3;
})(DevtoolsMainView || {});
var DEFAULT_MAIN_VIEW = "structure" /* Structure */;
var DebuggerModule = /* @__PURE__ */ ((DebuggerModule3) => {
  DebuggerModule3["Locator"] = "locator";
  DebuggerModule3["Structure"] = "structure";
  DebuggerModule3["Dgraph"] = "dgraph";
  return DebuggerModule3;
})(DebuggerModule || {});
var TreeWalkerMode = /* @__PURE__ */ ((TreeWalkerMode3) => {
  TreeWalkerMode3["Owners"] = "owners";
  TreeWalkerMode3["Components"] = "components";
  TreeWalkerMode3["DOM"] = "dom";
  return TreeWalkerMode3;
})(TreeWalkerMode || {});
var DEFAULT_WALKER_MODE = "components" /* Components */;
var NodeType = /* @__PURE__ */ ((NodeType3) => {
  NodeType3["Root"] = "ROOT";
  NodeType3["Component"] = "COMPONENT";
  NodeType3["Element"] = "ELEMENT";
  NodeType3["Effect"] = "EFFECT";
  NodeType3["Render"] = "RENDER";
  NodeType3["Memo"] = "MEMO";
  NodeType3["Computation"] = "COMPUTATION";
  NodeType3["Refresh"] = "REFRESH";
  NodeType3["Context"] = "CONTEXT";
  NodeType3["CatchError"] = "CATCH_ERROR";
  NodeType3["Signal"] = "SIGNAL";
  NodeType3["Store"] = "STORE";
  NodeType3["CustomValue"] = "CUSTOM_VALUE";
  return NodeType3;
})(NodeType || {});
var NODE_TYPE_NAMES = {
  ["ROOT" /* Root */]: "Root",
  ["COMPONENT" /* Component */]: "Component",
  ["ELEMENT" /* Element */]: "Element",
  ["EFFECT" /* Effect */]: "Effect",
  ["RENDER" /* Render */]: "Render Effect",
  ["MEMO" /* Memo */]: "Memo",
  ["COMPUTATION" /* Computation */]: "Computation",
  ["REFRESH" /* Refresh */]: "Refresh",
  ["CONTEXT" /* Context */]: "Context",
  ["CATCH_ERROR" /* CatchError */]: "CatchError",
  ["SIGNAL" /* Signal */]: "Signal",
  ["STORE" /* Store */]: "Store",
  ["CUSTOM_VALUE" /* CustomValue */]: "Custom Value"
};
var ValueItemType = /* @__PURE__ */ ((ValueItemType2) => {
  ValueItemType2["Signal"] = "signal";
  ValueItemType2["Prop"] = "prop";
  ValueItemType2["Value"] = "value";
  return ValueItemType2;
})(ValueItemType || {});
var UNKNOWN = "unknown";
var OWNER_LOCATION_PROP = "sdtLocation";

// packages/debugger/src/main/types.ts
var INSPECTED_STATE_NULL = {
  ownerId: null,
  signalId: null,
  treeWalkerOwnerId: null
};
var getValueItemId = (type, id) => {
  if (type === "value" /* Value */) return "value" /* Value */;
  return `${type}:${id}`;
};

export {
  DevtoolsMainView,
  DEFAULT_MAIN_VIEW,
  DebuggerModule,
  TreeWalkerMode,
  DEFAULT_WALKER_MODE,
  NodeType,
  NODE_TYPE_NAMES,
  ValueItemType,
  UNKNOWN,
  OWNER_LOCATION_PROP,
  INFINITY,
  NEGATIVE_INFINITY,
  NAN,
  UNDEFINED,
  ValueType,
  PropGetterState,
  GLOBAL_GET_VALUE,
  WINDOW_PROJECTPATH_PROPERTY,
  LOCATION_ATTRIBUTE_NAME,
  INSPECTED_STATE_NULL,
  getValueItemId
};
