// @solid-devtools/debugger/dist/setup.js
// packages/debugger/src/setup.ts
import * as s from "solid-js";
import * as store from "solid-js/store";
import { error } from "@solid-devtools/shared/utils";
var PassedLocatorOptions = null;
function useLocator(options) {
  PassedLocatorOptions = options;
}
function setLocatorOptions(options) {
  PassedLocatorOptions = options;
}
var ClientVersion = null;
var SolidVersion = null;
var ExpectedSolidVersion = null;
function setClientVersion(version) {
  ClientVersion = version;
}
function setSolidVersion(version, expected) {
  SolidVersion = version;
  ExpectedSolidVersion = expected;
}
if (globalThis.SolidDevtools$$) {
  error("Debugger is already setup");
}
if (!s.DEV || !store.DEV) {
  error("SolidJS in not in development mode!");
} else {
  let created_owners = [];
  let setup = {
    solid: {
      ...s.DEV,
      getOwner: s.getOwner,
      getListener: s.getListener,
      untrack: s.untrack,
      $PROXY: s.$PROXY,
      $TRACK: s.$TRACK,
      $DEVCOMP: s.$DEVCOMP,
      sharedConfig: s.sharedConfig
    },
    store: {
      ...store.DEV,
      unwrap: store.unwrap,
      $RAW: store.$RAW
    },
    get_created_owners() {
      const events = created_owners ?? [];
      created_owners = null;
      return events;
    },
    get_locator_options() {
      return PassedLocatorOptions;
    },
    versions: {
      get_client() {
        return ClientVersion;
      },
      get_solid() {
        return SolidVersion;
      },
      get_expected_solid() {
        return ExpectedSolidVersion;
      }
    },
    unowned: {
      signals: [],
      onSignalAdded: null,
      onSignalRemoved: null
    }
  };
  globalThis.SolidDevtools$$ = setup;
  s.DEV.hooks.afterCreateOwner = (owner) => {
    created_owners?.push(owner);
  };
  let signals_registry = new FinalizationRegistry((ref) => {
    let idx = setup.unowned.signals.indexOf(ref);
    setup.unowned.signals.splice(idx, 1);
    setup.unowned.onSignalRemoved?.(ref, idx);
  });
  s.DEV.hooks.afterCreateSignal = (signal) => {
    if (s.getOwner() == null) {
      let ref = new WeakRef(signal);
      let idx = setup.unowned.signals.push(ref) - 1;
      signals_registry.register(signal, ref);
      setup.unowned.onSignalAdded?.(ref, idx);
    }
  };
}
export {
  setClientVersion,
  setLocatorOptions,
  setSolidVersion,
  useLocator
};
